### PR TITLE
support loading from memory

### PR DESCRIPTION
**Pull request by luotao1**
support load program and parameters from memory.
usage:
```
cfg->SetProgBufferAndParamBuffer(&buffer_prog[0], buffer_prog.size(), &buffer_param[0], buffer_param.size());
```

### CODE REVIEW

File: paddle/fluid/framework/executor_thread_worker.cc [hunk 1]
```diff
@@ -97,7 +97,7 @@ void ExecutorThreadWorker::SetDevice() {
   static unsigned concurrency_cap = std::thread::hardware_concurrency();
   int thread_id = this->thread_id_;
 
-  if (thread_id < concurrency_cap) {
+  if ((unsigned)thread_id < concurrency_cap) {
```
Comment (Superjomn) [inline paddle/fluid/framework/executor_thread_worker.cc]: use static_cast instead, the (T) is the reintrpret_cast, which is very dangerous without testing

File: paddle/fluid/inference/io.cc [hunk 2]
```diff
@@ -130,27 +132,42 @@ std::unique_ptr<framework::ProgramDesc> Load(framework::Executor* executor,
...
+  LoadPersistables(executor, scope, *main_program, dirname, "",
+                   false /* is_memory_load */);
   return main_program;
 }
 
-std::unique_ptr<framework::ProgramDesc> Load(
-    framework::Executor* executor, framework::Scope* scope,
-    const std::string& prog_filename, const std::string& param_filename) {
-  std::string model_filename = prog_filename;
+std::unique_ptr<framework::ProgramDesc> Load(framework::Executor* executor,
```
Comment (Superjomn) [inline paddle/fluid/inference/io.cc]: Better to split the old one with the new one loading from memory, it is very tricky to make `prog_name` store the program buffer.

Just use an overloading function is more clear.

File: paddle/fluid/inference/io.cc [hunk 3]
```diff
@@ -130,27 +132,42 @@ std::unique_ptr<framework::ProgramDesc> Load(framework::Executor* executor,
...
+                                             framework::Scope* scope,
+                                             const std::string& prog_filename,
+                                             const std::string& param_filename,
+                                             bool is_memory_load = false) {
   std::string program_desc_str;
-  ReadBinaryFile(model_filename, &program_desc_str);
+  if (!is_memory_load) {
+    ReadBinaryFile(prog_filename, &program_desc_str);
+  } else {
+    program_desc_str = prog_filename;
```
Comment (Superjomn) [inline paddle/fluid/inference/io.cc]: here is tricky.

File: paddle/fluid/inference/io.cc [hunk 4]
```diff
@@ -130,27 +132,42 @@ std::unique_ptr<framework::ProgramDesc> Load(framework::Executor* executor,
...
+  }
 
   std::unique_ptr<framework::ProgramDesc> main_program(
       new framework::ProgramDesc(program_desc_str));
   PADDLE_ENFORCE(framework::IsProgramVersionSupported(main_program->Version()),
                  "model version %ld is not supported.",
                  main_program->Version());
 
-  LoadPersistables(executor, scope, *main_program, "", param_filename);
+  LoadPersistables(executor, scope, *main_program, "", param_filename,
```
Comment (Superjomn) [inline paddle/fluid/inference/io.cc:153]: Use an overloadding function instread.

```c++
if (is_memory_load) {
  LoadXXX(...); // load from memory
} else {
  LoadXXX(...);
}
```

File: paddle/fluid/inference/io.h [hunk 5]
```diff
@@ -41,6 +41,12 @@ std::unique_ptr<framework::ProgramDesc> Load(framework::Executor* executor,
                                              const std::string& prog_filename,
                                              const std::string& param_filename);
 
+std::unique_ptr<framework::ProgramDesc> Load(framework::Executor* executor,
```
Comment (Superjomn) [inline paddle/fluid/inference/io.h]: Add an overloading function instead change the old one.

File: paddle/fluid/inference/api/analysis_config.cc [hunk 6]
```diff
@@ -102,4 +105,12 @@ void contrib::AnalysisConfig::EnableTensorRtEngine(int workspace_size,
   pass_builder()->InsertPass(1, "tensorrt_subgraph_pass");
 }
 
+void contrib::AnalysisConfig::SetProgBufferAndParamBuffer(
```
Comment (Superjomn) [inline paddle/fluid/inference/api/analysis_config.cc]: `SetModelBuffer` seems shorter and clear.

File: paddle/fluid/inference/api/analysis_predictor.cc [hunk 7]
```diff
@@ -448,27 +449,31 @@ bool AnalysisPredictor::LoadProgramDesc() {
...
-  fin.close();
-
   // Create ProgramDesc
   framework::proto::ProgramDesc proto;
-  proto.ParseFromString(pb_content);
+  if (!config_.is_memory_load()) {
+    std::string pb_content;
+    // Read binary
+    std::ifstream fin(filename, std::ios::in | std::ios::binary);
+    PADDLE_ENFORCE(static_cast<bool>(fin), "Cannot open file %s", filename);
```
Comment (Superjomn) [inline paddle/fluid/inference/api/analysis_predictor.cc]: `PADDLE_ENFORCE(fin.is_open())`

File: paddle/fluid/inference/api/paddle_analysis_config.h [hunk 8]
```diff
@@ -52,10 +52,15 @@ struct AnalysisConfig : public NativeConfig {
...
-  // NOTE this is just for internal development, please not use it.
-  // NOT stable yet.
   bool use_mkldnn() const { return use_mkldnn_; }
 
+  // Specify the memory buffer of program and parameter
+  void SetProgBufferAndParamBuffer(const char* prog_buffer,
+                                   size_t prog_buffer_size,
+                                   const char* program_buffer,
+                                   size_t program_buffer_size);
+  bool is_memory_load() const { return is_memory_load_; }
```
Comment (Superjomn) [inline paddle/fluid/inference/api/paddle_analysis_config.h]: `model_from_memory`

File: paddle/fluid/framework/executor_thread_worker.cc (hunk 1 above)
Comment (luotao1) [reply paddle/fluid/framework/executor_thread_worker.cc]: Done

File: paddle/fluid/inference/api/analysis_config.cc (hunk 6 above)
Comment (luotao1) [reply paddle/fluid/inference/api/analysis_config.cc]: Done

File: paddle/fluid/inference/api/analysis_predictor.cc (hunk 7 above)
Comment (luotao1) [reply paddle/fluid/inference/api/analysis_predictor.cc]: Done

File: paddle/fluid/inference/api/paddle_analysis_config.h (hunk 8 above)
Comment (luotao1) [reply paddle/fluid/inference/api/paddle_analysis_config.h]: Done

File: paddle/fluid/inference/io.cc (hunk 2 above)
Comment (luotao1) [reply paddle/fluid/inference/io.cc]: Create a `LoadFromMemory` function. 
```
std::unique_ptr<framework::ProgramDesc> LoadFromMemory(
framework::Executor* executor,  
framework::Scope* scope,
const std::string& prog_buffer, 
const std::string& param_buffer)
```
The reason why not use an overloading function is that the parameter type and counts in new `LoadFromMemory` function are the same with the old `Load` function.

File: paddle/fluid/inference/io.cc (hunk 3 above)
Comment (luotao1) [reply paddle/fluid/inference/io.cc]: Done

File: paddle/fluid/inference/io.cc (hunk 4 above)
Comment (luotao1) [reply paddle/fluid/inference/io.cc:153]: Done.

File: paddle/fluid/inference/io.h (hunk 5 above)
Comment (luotao1) [reply paddle/fluid/inference/io.h]: Create a `LoadFromMemory` function.

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**